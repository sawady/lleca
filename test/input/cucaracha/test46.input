
fun fib(n : Int) : Int {
    a := 0
    b := 1
    i := 0
    while i < n {
        s := a + b
        a := b
        b := s
        i := i + 1
    }
    return a
}

fun main() {
  i := 0
  while i <= 20 {
    putNum(fib(i))
    putChar(10)
    i := i + 1
  }
}

