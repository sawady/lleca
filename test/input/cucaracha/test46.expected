Cons(
  Function(
    fib,
    Cons(
      Parameter(
        n,
        Int
      ),
      Nil
    ),
    Int,
    Cons(
      StmtAssign(
        a,
        ExprConstNum(
          0
        )
      ),
      Cons(
        StmtAssign(
          b,
          ExprConstNum(
            1
          )
        ),
        Cons(
          StmtAssign(
            i,
            ExprConstNum(
              0
            )
          ),
          Cons(
            StmtWhile(
              ExprLt(
                ExprVar(
                  i
                ),
                ExprVar(
                  n
                )
              ),
              Cons(
                StmtAssign(
                  s,
                  ExprAdd(
                    ExprVar(
                      a
                    ),
                    ExprVar(
                      b
                    )
                  )
                ),
                Cons(
                  StmtAssign(
                    a,
                    ExprVar(
                      b
                    )
                  ),
                  Cons(
                    StmtAssign(
                      b,
                      ExprVar(
                        s
                      )
                    ),
                    Cons(
                      StmtAssign(
                        i,
                        ExprAdd(
                          ExprVar(
                            i
                          ),
                          ExprConstNum(
                            1
                          )
                        )
                      ),
                      Nil
                    )
                  )
                )
              )
            ),
            Cons(
              StmtReturn(
                ExprVar(
                  a
                )
              ),
              Nil
            )
          )
        )
      )
    )
  ),
  Cons(
    Function(
      main,
      Nil,
      Unit,
      Cons(
        StmtAssign(
          i,
          ExprConstNum(
            0
          )
        ),
        Cons(
          StmtWhile(
            ExprLe(
              ExprVar(
                i
              ),
              ExprConstNum(
                20
              )
            ),
            Cons(
              StmtCall(
                putNum,
                Cons(
                  ExprCall(
                    fib,
                    Cons(
                      ExprVar(
                        i
                      ),
                      Nil
                    )
                  ),
                  Nil
                )
              ),
              Cons(
                StmtCall(
                  putChar,
                  Cons(
                    ExprConstNum(
                      10
                    ),
                    Nil
                  )
                ),
                Cons(
                  StmtAssign(
                    i,
                    ExprAdd(
                      ExprVar(
                        i
                      ),
                      ExprConstNum(
                        1
                      )
                    )
                  ),
                  Nil
                )
              )
            )
          ),
          Nil
        )
      )
    ),
    Nil
  )
)
